core/src/rules/optimising_line_formatter/mod.rs

```diff
@@ -115,6 +104,71 @@ impl OptimisingLineFormatter {
             recon_settings,
         }
     }
+
+    /// Creates a map between tokens and their child lines.
+    ///
+    /// A line's direct parent doesn't necessarily lead to a sequential
+    /// exploration of tokens/child lines. To fix this, the `LineParent` is
+    /// mapped to the nearest previous token on the parent line.
+    fn get_line_children(lines: &[LogicalLine]) -> FxHashMap<LineParent, LineChildren> {
+        let mut tokens_before_gaps = FxHashMap::<usize, Vec<usize>>::default();
+        let mut line_parent_map = FxHashMap::<LineParent, LineParent>::default();
+        let mut line_children = FxHashMap::<LineParent, LineChildren>::default();
+
+        let get_line_tokens_before_gaps = |line_index: usize| {
+            let line_tokens_iter = lines[line_index].get_tokens().iter();
+            // Add extra element so last token can been seen to have a "gap" after it
+            let line_tokens_iter = line_tokens_iter.chain(once(&usize::MAX));
+            line_tokens_iter
+                .fold((None, vec![]), |(last_index, mut tokens), &token_index| {
+                    tokens.extend(last_index.filter(|&last_index| last_index + 1 != token_index));
+                    (Some(token_index), tokens)
+                })
+                .1
+        };
+
+        for (line_index, line) in lines.iter().enumerate() {
+            let Some(&first_token_index) = line.get_tokens().first() else {
+                continue;
+            };
+            if let Some(parent) = line.get_parent() {
+                let expected_parent_token = (|| {
+                    let gap_tokens = tokens_before_gaps
+                        .entry(parent.line_index)
+                        .or_insert_with_key(|&line_index| get_line_tokens_before_gaps(line_index));
+                    let gap_index = gap_tokens
+                        .partition_point(|&gap| gap < first_token_index)
+                        .checked_sub(1)?;
+                    gap_tokens.get(gap_index).cloned()
+                })();
+
+                if let (Some(expected_parent_token), Entry::Vacant(entry)) = (
+                    expected_parent_token.filter(|&idx| idx != parent.global_token_index),
+                    line_parent_map.entry(parent),
+                ) {
+                    entry.insert(LineParent {
+                        line_index: parent.line_index,
+                        global_token_index: expected_parent_token,
+                    });
+                }
+            }
+
+            let mut current_line = Some(line);
+            let mut first_parent = true;
+            while let Some(parent) = current_line.and_then(LogicalLine::get_parent) {
+                let token_children = line_children
```

Comment:
with the map on the right renamed to `line_children_map` (or something similar), the left variable should probably be `line_children`